### ISSUE TITLE
Rule proposal: Don't pass default values to functions

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#13168]
**Please describe what the rule should do:**

Warn about using function parameters that are already set by the function's defaults.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[x] Other (please specify:) **_Reduces code_**

**Provide 2-3 code examples that this rule will warn about:**

```js
function run(always = true) {}
run(true); // Error
```
```js
function run({always} = {always: true}) {}
run({always: true}); // Error
```
```js
function run({always = true} = {}) {}
run({always: true}); // Error
```
```js
function run(options) {
	options = {
		always: true,
		...options
	}
}
run({always: true}); // Error
```
```js
function run(options) {
	options = Object.assign({
		always: true
	}, options);
}
run({always: true}); // Error
```

And any other common defaults pattern (unless already blocked by other rules like [`prefer-object-spread`](https://eslint.org/docs/rules/prefer-object-spread))

**Why should this rule be included in ESLint (instead of a plugin)?**

I think this is something you should tell me, I don't know about your reasoning about what belongs to ESLint and what doesn't. Could you include that in GitHub's Issue template?

**Comment by ljharb** [comment]
eslint generally doesn't run across files, so the only way this is useful is if the function being called is declared in the same file - which means it's not particularly useful.

**Comment by ljharb** [comment]
Separately, overriding defaults is the *entire point* of default arguments, even when putting the same value. In your example, `run(true)` is actually *protecting* the caller from when `run` changes the default to `false`.

**Comment by fregante** [comment]
> eslint generally doesn't run across files, so the only way this is useful is if the function being called is declared in the same file - which means it's not particularly useful.

From my (little) understanding, TypeScript can access types across files. `eslint-plugin-import` looks inside the imported files as well. The utility of this rule only depends on the choices taken while developing it. Is that correct?

> `run(true)` is actually _protecting_ the caller

That I don't disagree with, but... (look at my example below)

> when `run` changes the default to `false`.

That's a breaking change. It breaks.

---

Let me make a more practical example. Which one do you prefer?

```js
fetch('/contacts');
```
```js
fetch('/contacts', {
	method: 'GET',
	headers: {},
	mode: 'cors',
	credentials: 'include',
	cache: 'default',
	redirect: 'follow'
});
```

They're equivalent. Neither one of us prefers the second version. This rule _prevents the second version._

**Comment by ljharb** [comment]
Yes, but this is eslint core, eslint-plugin-import only looks across files for exports/imports, and typescript isn't relevant here. The choice to look across files is not a small one, and I suspect it's not one eslint core rules will ever take.

Actually, you're wrong - I prefer the second version, strongly, because I find relying on implicit defaults dangerous, and lacking self-documentation.

**Comment by mdjermanovic** [comment]
I agree with [this comment](https://github.com/eslint/eslint/issues/13168#issuecomment-611785954) @ljharb, it looks like this rule wouldn't provide much value since it would be limited to function calls in the same file where the function is defined.

Users could reasonably expect that it also works on imported/global functions and might see a lack of warnings as false negatives.

**Comment by fregante** [comment]
> Actually, you're wrong - I prefer the second version, strongly

You forgot to specify the HTTP status here, then: https://github.com/ljharb/tc39-ipr-check/blob/cc15bb5e28e8d7cbb8ee2f6f6d244c72920f90db/src/http/get-index/index.js#L32-L40

Why didn't you set `statusCode: 200` if you prefer it? That's a default.

**Comment by ljharb** [comment]
"knowingly" is the issue. In that case, i didn't even write that code; it's what a generator spat out for the service it uses.

Finding counterexamples doesn't affect my *preference*, which is always to specify defaults.
